### PR TITLE
feat: Implement tool mode functionality and dynamic placeholders across input components

### PR DESCRIPTION
**Pull request by ogabrielluiz**
This pull request includes several changes to enhance the functionality and maintainability of the `langflow` backend. The most important changes include adding support for asynchronous output functions, improving tool mode handling, and updating constants and methods for better clarity and consistency.

### Enhancements to Asynchronous Functionality:
* Added an asynchronous output function builder `_build_output_async_function` to handle async methods in `component_tool.py`.
* Updated `get_tools` method to support asynchronous output methods and ensure all required inputs are in tool mode.

### Tool Mode Handling:
* Introduced `tool_mode` attribute to `MessageTextInput` in `agent.py` to handle specific tool mode inputs.
* Updated the `run_and_validate_update_outputs` method to replace outputs based on `tool_mode`.
* Added methods `_append_tool_to_outputs_map` and `_build_tool_output` to handle tool output mapping in `component.py` [[1]](diffhunk://#diff-22fd745de33f20981c7cbcbf3c91d4a6f8719de20761eb22cf42f8ed6ec50169R1044-R1049) [[2]](diffhunk://#diff-22fd745de33f20981c7cbcbf3c91d4a6f8719de20761eb22cf42f8ed6ec50169L903-R914).

### Constants and Methods Updates:
* Added `TOOL_OUTPUT_DISPLAY_NAME` constant in `constants.py` and updated its usage across the codebase [[1]](diffhunk://#diff-c60ffce8614e539a6fa7c3e47ac11aa7b89f15b9e09d9549c907ef15a7d33894R2) [[2]](diffhunk://#diff-22fd745de33f20981c7cbcbf3c91d4a6f8719de20761eb22cf42f8ed6ec50169L15-R15).
* Renamed `get_function_entrypoint_return_type` to `_get_function_entrypoint_return_type` for better encapsulation in `custom_component.py`.

### Code Cleanup:
* Removed unused import `FEATURE_FLAGS` in `component.py`.
* Simplified the initial setup JSON file by removing redundant code.

### CODE REVIEW

**Comment by edwinjosechittilappilly** [comment]
@ogabrielluiz Few Issues on testing the PR:

The add_tool_output seems to be not working. Example in the URL component, are we planning to have tool mode button on that component also?
Does the tool mode work in every component?

Also Simple Agent Template that we are loading by default seems not to have the tool mode enabling button, should we update the template?

**Comment by edwinjosechittilappilly** [comment]
> @ogabrielluiz Few Issues on testing the PR:
> 
> The add_tool_output seems to be not working. Example in the URL component, are we planning to have tool mode button on that component also? Does the tool mode work in every component?
> 
> Also Simple Agent Template that we are loading by default seems not to have the tool mode enabling button, should we update the template?

Found,
If we add tool_mode = True then we get Tool mode button at top of the component.
we might need to enable tool mode for certain components for better UX

good work.

**Comment by edwinjosechittilappilly** [comment]
Only Pending Error:

StatementError: (builtins.TypeError) Object of type datetime is not JSON serializable
                             [SQL: UPDATE message SET content_blocks=? WHERE message.id = ?]
                             [parameters: [{'content_blocks': [{'title': 'Agent Input', 'content': {'type': 'text', 'text': "Agent initiated with input: {'input': 'search
                             https://en.wikipedia.o ... (430434 characters truncated) ... n\n\n\n\n\n\n\n\n\n', 'files': []}}, 'allow_markdown': True}]}},
                             'allow_markdown': True}], 'message_id': UUID('2e3c1532-21fe-491e-8f0a-3b2d5b6d52ee')}]]

**Comment by edwinjosechittilappilly** [comment]
Error when execution in    Events.py
   tool_content = _find_or_create_tool_content(agent_message, tool_name, tool_input=tool_input)
 Error:
TypeError: _find_or_create_tool_content() got multiple values for argument 'tool_input'

possible fix:
tool_content = _find_or_create_tool_content(agent_message, tool_name, tool_input)

File: src/backend/base/langflow/base/tools/component_tool.py [hunk 1]
```diff
@@ -78,19 +174,56 @@ def get_tools(self) -> list[BaseTool]:
...
+                    StructuredTool(
+                        name=formatted_name,
+                        description=build_description(self.component, output),
+                        func=_build_output_function(self.component, output_method, event_manager),
+                        args_schema=args_schema,
+                        handle_tool_error=True,
+                        callbacks=callbacks,
+                    )
                 )
+        if len(tools) == 1:
```
Comment (edwinjosechittilappilly) [inline src/backend/base/langflow/base/tools/component_tool.py:219]: This Causes Errors when only one tool for component:
Tested Edit:
This change would remove the error in components as tools for custom component.

`if len(tools) == 1 and (tool_name and tool_description)`

**Review by edwinjosechittilappilly — CHANGES_REQUESTED** [review]

**Comment by edwinjosechittilappilly** [comment]
Agent Name and Description not updated when Agent is turned to a tool in tool mode. The Agent as tool always have the default name and description, which could be a blocker when calling it as a tool.

File: src/backend/base/langflow/custom/custom_component/component.py [hunk 2]
```diff
@@ -98,8 +97,6 @@ def __init__(self, **kwargs) -> None:
         self.__config = config
         self._reset_all_output_values()
         super().__init__(**config)
-        if (FEATURE_FLAGS.add_toolkit_output) and hasattr(self, "_append_tool_output") and self.add_tool_output:
-            self._append_tool_output()
```
Comment (phact) [inline src/backend/base/langflow/custom/custom_component/component.py:102]: this means we're not enabling components as tools in this PR? Just getting rid of the feature flag?

**Review by phact — APPROVED** [review]
I like the approach, tested it manually and really enjoy the playground view.

It's great that we show the inputs as part of the steps but we should have a way to fold them since they quickly get long (in a separate PR).

We should add more tests in a separate PR.

**Review by edwinjosechittilappilly — APPROVED** [review]
LGTM.
Tested

File: src/backend/base/langflow/custom/custom_component/component.py (hunk 2 above)
Comment (edwinjosechittilappilly) [reply src/backend/base/langflow/custom/custom_component/component.py:102]: To enable a component as  tools, we need to add tool_mode as true for relevant inputs of the component and we shall get the ToolMode button in component.
Screenshot 2024-11-07 at 5 11 31 PM

PR to add Tool mode to relevant compoenents : #4439

Comment (phact) [reply src/backend/base/langflow/custom/custom_component/component.py:102]: Gotcha thanks!

**Pull request merged**